### ISSUE TITLE
Propose libtorrent/BEP "BTShare" - websocket API for torrent clients

### ISSUE THREAD

**Issue by SoniEx2** [Chocobozzz/PeerTube#2323]
This is gonna be a tricky one to talk about, but let's give it a try.

So, I think peertube should draft up a BEP for torrent clients to implement, that allows peertube (or other web apps, or even minecraft mods) to open a websocket to `ws://127.0.0.1:7087/btshare`. This websocket would then give you a storageless torrent client - instead of having storage attached, you'd write the storage handling yourself. This is particularly useful for peertube because we don't want to use the filesystem. On the other hand minecraft mods would benefit from being able to use the filesystem.

Additionally, this follows the torrent client's existing limits. (Or additional limits could be configured for separate websocket clients, either based on "Origin" or otherwise.) This is good for the user, and is even better for the user if ppl start embedding bittorrent clients into stuff (which is something I definitely wanna do in the near future). This way, those clients can detect the websocket API and opt to use it rather than doing their own peer management and stuff. (This would also disable peer management settings as that would be handled by the main torrent client.)

All in all this seems to benefit everyone, from peertube users, to embedded bittorrent users, and probably others too.

(And yeah noScript breaks this. But if you're using noScript you're supposed to know how to add an exception for it - as you'd already have to do for peertube itself, which uses WebRTC which is blocked by default.)

**Comment by rigelk** [comment]
hi @SoniEx2 ! Thanks for the proposal. I'm afraid I understood little of it however. I'm trying to clarify, so please bear with me:

From my understanding, you consider PeerTube a torrent client in your scenario (like other web apps or as you cite, minecraft mods), right? That client would then open a connection to the suggested endpoint (`ws://127.0.0.1:7087/btshare`) to store/read data, a thing left to the implementer of the endpoint.

However, PeerTube is not just a web app. It is also a web server. What part of it would you consider to be the client in your scenario?

---

### Questions

> This websocket would then give you a storageless torrent client […] This is particularly useful for peertube because we don't want to use the filesystem.

PeerTube's server stores videos on disk, which is a limited resource. PeerTube's web client stores video chunks received from the server and peers in its browser storage (IndexedDB), which is much less limited as it is present in memory/disk for a limited time.

> Additionally, this follows the torrent client's existing limits. (Or additional limits could be configured for separate websocket clients, either based on "Origin" or otherwise.) This is good for the user, and is even better for the user if ppl start embedding bittorrent clients into stuff.

That is wher I don't understand. Limits exist on traditional clients since they usually seed multiple torrents at the same time. PeerTube as a web app seeds one torrent at a time (the video being watched), so it doesn't apply to it. I assume you are referring to other implementations of your BEP, but I don't see how this concerns PeerTube.

> This way, those clients can detect the websocket API and opt to use it rather than doing their own peer management and stuff. (This would also disable peer management settings as that would be handled by the main torrent client.)

I thought the suggested API was a storage API, not a management one.

---

Related to your suggestion, I think the torrent community at large would benefit from having a BEP for WebTorrent (which PeerTube uses for its torrent client), which right now is more a _de facto_ standard than a real BitTorrent standard. (specifically, https://github.com/webtorrent/webtorrent/pull/881). Pushing and helping the implementation at https://github.com/arvidn/libtorrent/issues/223 might help too.

**Comment by SoniEx2** [comment]
okay, let me try again. (hopefully I can get it right this time.)

- - -

For starters, I don't want a BEP for webtorrent, instead I want a BEP to enable websites to use locally-installed torrent clients.

Peertube, or at least the "client-side javascript" of peertube, would open a connection to `ws://127.0.0.1:7087/btshare`. This would then allow it to queue up one or more torrent/magnet links. The daemon running on that endpoint would then ask the websocket client (Peertube's client-side javascript) for how many pieces it has available (if any). Then it would start fetching that torrent from other peers, as well as seeding to other peers.

When receiving pieces, it would forward them to the websocket client to store, as if the websocket client was a filesystem, or had one available. Equivalently, when seeding, it would ask the websocket client to give it some pieces, and forward those to the remote peer. (This is important because the API should be sandboxed, and this is meant to help achieve that.)

- - - 

The suggested API is a heavily sandboxed API with two related functions:

1. Allows a "primary" torrent client, installed by the user, to take on the task of managing torrents and setting limits, with "secondary" torrent clients, such as ones embedded into other programs, relying on the "primary" torrent client for peer-to-peer communications.
2. Enables web apps (javascript running on the user's web browser) to freely use bittorrent, within limits specified by the user's "primary" torrent client. The web apps act as "secondary" torrent clients.

It's also meant to suit another goal, which is to not fragment the bittorrent protocol. WebTorrents don't support DHT, UDP trackers or uTP, and my proposal enables web apps to benefit from DHT, UDP trackers and uTP, and all of the advantages that come with them.

**Comment by Nutomic** [comment]
Sorry but what does this have to do with Peertube? Sounds like it should be a completely seperate project.

**Comment by SoniEx2** [comment]
It has to do with Peertube using it.

As such, the API needs to be usable by peertube.

**Comment by Chocobozzz** [comment]
Hello,

Sorry but it's out of the scope of the PeerTube project.
